sdks/python/apache_beam/transforms/enrichment_handlers/vertex_ai_feature_store.py

```diff
@@ -0,0 +1,289 @@
+    try:
+      # checks if feature store exists
+      _ = aiplatform.Featurestore(
+          featurestore_name=self.feature_store_id,
+          project=self.project,
+          location=self.location,
+          credentials=self.kwargs.get('credentials'),
+      )
+      if 'client_options' in self.kwargs:
+        self.client = aiplatform.gapic.FeaturestoreOnlineServingServiceClient(
+            **self.kwargs)
+      else:
+        self.client = aiplatform.gapic.FeaturestoreOnlineServingServiceClient(
+            client_options={'api_endpoint': self.api_endpoint}, **self.kwargs)
+      self.entity_type_path = self.client.entity_type_path(
+          self.project,
+          self.location,
+          self.feature_store_id,
+          self.entity_type_id)
+    except NotFound:
+      raise ValueError(
+          'Vertex AI Feature Store %s does not exist' % self.feature_store_id)
+
+  def __call__(self, request: beam.Row, *args, **kwargs):
+    """Fetches feature value for an entity-id from
+    Vertex AI Feature Store (Legacy).
+
+    Args:
+      request: the input `beam.Row` to enrich.
+    """
+    try:
+      entity_id = request._asdict()[self.row_key]
+    except KeyError:
+      raise ValueError(
+          "no entry found for row_key %s in input row" % self.row_key)
+
+    try:
+      selector = aiplatform.gapic.FeatureSelector(
+          id_matcher=aiplatform.gapic.IdMatcher(ids=self.feature_ids))
+      response = self.client.read_feature_values(
+          request=aiplatform.gapic.ReadFeatureValuesRequest(
+              entity_type=self.entity_type_path,
+              entity_id=entity_id,
+              feature_selector=selector))
+    except NotFound:
+      raise ValueError(
+          _not_found_err_message(
+              self.feature_store_id, self.entity_type_id, entity_id))
+
+    response_dict = {}
+    proto_to_dict = proto.Message.to_dict(response.entity_view)
+    for key, msg in zip(response.header.feature_descriptors,
+                        proto_to_dict['data']):
+      if msg and 'value' in msg:
+        for _, value in msg['value'].items():
+          response_dict[key.id] = value
+          break  # skip fetching the metadata
+      elif self.exception_level == ExceptionLevel.RAISE:
+        raise ValueError(
+            _not_found_err_message(
+                self.feature_store_id, self.entity_type_id, entity_id))
+      elif self.exception_level == ExceptionLevel.WARN:
+        _LOGGER.warning(
+            _not_found_err_message(
+                self.feature_store_id, self.entity_type_id, entity_id))
+    return request, beam.Row(**response_dict)
+
+  def __exit__(self, exc_type, exc_val, exc_tb):
+    """Clean the instantiated Vertex AI client."""
+    self.client = None
+
+  def get_cache_key(self, request: beam.Row) -> str:
+    """Returns a string formatted with unique entity-id for the feature values.
+    """
+    return 'entity_id: %s'
```

Comment:
Same as above, should we be grabbing entity_id from the request?

Reply:
good catch, IDK how I missed this